### ISSUE TITLE
RTOS environment

### ISSUE THREAD

**Issue by troky** [BlueAndi/vscp-framework#30]
I am using vscp library in RTOS environment in its own task where I call vscp_core_process() in endless loop.
CAN messages are received and stored in FIFO queue in ISR because I don't want to lose any message.
vscp_tp_adapter_readMessage() reads messages from above FIFO, not directly.

Everything works well even in processing local loop messages.

I have some situations when I need to send 50-200 zoned stream data messages in a row without waiting for confirmation because performance is very important.
Problem is when I receive more messages than I have space in FIFO because vscp_core_process() is slow consumer.

I have several other tasks and limited RAM resources so I can't increase vscpTask priority nor I can increase FIFO size (currently 64).

I played with task priorites and got working solution like this:
- set vscpTask high priority
```
vscpTask() {
  while(1) {
	...
	do {
	  kick_watchdog();
	} while (vscp_core_process()); // consume rx messages as fast as possible
	osDelay(5); // give other lower priority tasks some time
	...
  }
}
```
... of course, I had to change vscp_core_process() to return TRUE when valid message is received, FALSE otherwise.

Any other ideas how to solve the problem?
Is it possible to change vscp_core_process() as discribed above?

**Comment by BlueAndi** [comment]
Hmm ... here are some things, which came up very quick in my mind:

The vscp core must be kicked in case a message is available or caused by a timer timeout. To gain performance the vscpTask() should wait on a os signal (element in queue or timer has timeout), using osWait() with infinite wait time. This could make it a little bit faster.

Do you use the vscp transport queue too to loop sent events through the decision matrix? I guess not, isn't it?

Do you push whole events to queue or only pointers to event locations (heap or static)? I think about a "zero-copy" strategy, but I guess you already considered this.

Would it be possible to make a pre-filter in the ISR to avoid pushing every message to the queue? I am not sure whether every event is for the node. But of course it may be complex depended on the decision matrix configuration. Just an idea.

And yes, we can add a return value to vscp_core_process(). Lets first figure out all possibilities.

**Comment by troky** [comment]
> The vscp core must be kicked in case a message is available or caused by a timer timeout. To gain performance the vscpTask() should wait on a os signal (element in queue or timer has timeout), using osWait() with infinite wait time. This could make it a little bit faster.

I could trigger signal, but vscp_core_process() has to be called even if there are no messages in the queue because of internal timers, right?
 
> Do you use the vscp transport queue too to loop sent events through the decision matrix? I guess not, isn't it?

Actually, yes. I have VSCP_CONFIG_ENABLE_LOOPBACK enabled and I run local TX messages through DM.

> Do you push whole events to queue or only pointers to event locations (heap or static)? I think about a "zero-copy" strategy, but I guess you already considered this.

In fact, I am using FIFO in CAN driver (not VSCP) where I put complete CAN frame in queue. It is completely separated from VSCP framework.
In _vscp_tp_adapter_readMessage()_ I call something like _CAN_ReceiveFromQueue()_ without wait timeout... so it returns immediately.

This mechanism works wery well so far.

> Would it be possible to make a pre-filter in the ISR to avoid pushing every message to the queue? I am not sure whether every event is for the node. But of course it may be complex depended on the decision matrix configuration. Just an idea.

Not possible.

> And yes, we can add a return value to vscp_core_process(). Lets first figure out all possibilities.

I see this as temporary workaround, not final solution, but it could be useful to implement anyway.

I think that solution would be to separate somehow (if possible) vscp message processing and handling internal timers. RTOS has own timers than can be useful and message processing could be simple queue consumer with osWaitForever.
Currently _vscp_timer_process()_ doesn't trigger anything except it only processes timers.

So, maybe we could make event driven (RTOS friendly) version of vscp-framework?

I am using FreeRTOS but only CMSIS-OS v1 compatible functions.

**Comment by BlueAndi** [comment]
> > The vscp core must be kicked in case a message is available or caused by a timer timeout. To gain performance the vscpTask() should wait on a os signal (element in queue or timer has timeout), using osWait() with infinite wait time. This could make it a little bit faster.
> 
> I could trigger signal, but vscp_core_process() has to be called even if there are no messages in the queue because of internal timers, right?

Yes, thats why I suggested to use osWait(). You can wait for a onTimer and onQueue event or more events. osWait() may have a different name at your side, I am not sure.

> 
> > Do you use the vscp transport queue too to loop sent events through the decision matrix? I guess not, isn't it?
> 
> Actually, yes. I have VSCP_CONFIG_ENABLE_LOOPBACK enabled and I run local TX messages through DM.

Ok.

> 
> > Do you push whole events to queue or only pointers to event locations (heap or static)? I think about a "zero-copy" strategy, but I guess you already considered this.
> 
> In fact, I am using FIFO in CAN driver (not VSCP) where I put complete CAN frame in queue. It is completely separated from VSCP framework.
> In _vscp_tp_adapter_readMessage()_ I call something like _CAN_ReceiveFromQueue()_ without wait timeout... so it returns immediately.
> 
> This mechanism works wery well so far.

Ok, I unterstand. Do you have only VSCP messages on the CAN bus?

> > Would it be possible to make a pre-filter in the ISR to avoid pushing every message to the queue? I am not sure whether every event is for the node. But of course it may be complex depended on the decision matrix configuration. Just an idea.
> 
> Not possible.

Is it possible to use filters on CAN transceiver side?

> 
> > And yes, we can add a return value to vscp_core_process(). Lets first figure out all possibilities.
> 
> I see this as temporary workaround, not final solution, but it could be useful to implement anyway.
> 
> I think that solution would be to separate somehow (if possible) vscp message processing and handling internal timers. RTOS has own timers than can be useful and message processing could be simple queue consumer with osWaitForever.
> Currently _vscp_timer_process()_ doesn't trigger anything except it only processes timers.
> 
> So, maybe we could make event driven (RTOS friendly) version of vscp-framework?
> 
> I am using FreeRTOS but only CMSIS-OS v1 compatible functions.

I think this is already possible, as I suggested in the first lines. You can assign os timers in vscp_timer_create(). Because the os timers are assigned to the vscp task, osWait() should be automatically triggered.

Can you point to the FreeRTOS CMSIS-OS v1 functions?

**Comment by troky** [comment]
Unfortunately, osWait() is not implemented in Freertos CMSIS wrappers even though CMSIS defines it.
https://www.keil.com/pack/doc/CMSIS/RTOS/html/group__CMSIS__RTOS__Wait.html

Yes, I have only VSCP messages on CAN bus and I can use filters, but why?

So... I can't do proposed solution. In fact, I can't even send signal when data is ready because queueing is done in CAN driver that is completely unaware od higher levels like vscp and I would like to keep it that (simple) way.

I think making this framework fully RTOS friendly requires too much hassle and code changing to separate timer and message processing and exposing some callbacks to outside world... or maybe you have some cool idea I don't see :)

So, I have working solution like this:
```
vscpTask() {
  ...
  while(1) {
    if (vscp_core_process()) {
      osThreadSetPriority(vscpTaskHandle, osPriorityAboveNormal);
      do {
        RTOSWATCHDOG_Kick(WATCHDOG_VSCP);
       } while (vscp_core_process());
       osThreadSetPriority(vscpTaskHandle, defaultPriority);
    }
    osDelay(1);
  }
}
```
Basically, when vscp message is received I increase task priority so it can consume all messages as fast as it can and then decrease proiorty to normal so timers/states can be processed inside _vscp_core_process()_ as usual.

_vscp_timer_process(100);_ is processed in RTOS timer callback every 100ms in separate thread.

just for the reference, I have this in _vscp_tp_adapter.c_ :
```
extern BOOL vscp_tp_adapter_readMessage(vscp_RxMessage *const msg) {
  ...
  HALP_CAN_Status_t status = HALP_CAN_ReceiveFromQueue(&hCAN, 0);
  ...
```
So I don't block _vscp_core_process()_ at all... but putting _osWaitForever_ instead _0_ it would block until message is received and with current implementation we can't do that because timers could expire but won't be processed in state machine.

**Comment by BlueAndi** [comment]
> Unfortunately, osWait() is not implemented in Freertos CMSIS wrappers even though CMSIS defines it.
> https://www.keil.com/pack/doc/CMSIS/RTOS/html/group__CMSIS__RTOS__Wait.html

*grrr* :-)


> Yes, I have only VSCP messages on CAN bus and I can use filters, but why?

I asked to get picture about how to reduce the incoming data stream.


> So... I can't do proposed solution. In fact, I can't even send signal when data is ready because queueing is done in CAN driver that is completely unaware od higher levels like vscp and I would like to keep it that (simple) way.

Ok.

> I think making this framework fully RTOS friendly requires too much hassle and code changing to separate timer and message processing and exposing some callbacks to outside world... or maybe you have some cool idea I don't see :)

:-)


> So, I have working solution like this:
> 
> ```
> vscpTask() {
>   ...
>   while(1) {
>     if (vscp_core_process()) {
>       osThreadSetPriority(vscpTaskHandle, osPriorityAboveNormal);
>       do {
>         RTOSWATCHDOG_Kick(WATCHDOG_VSCP);
>        } while (vscp_core_process());
>        osThreadSetPriority(vscpTaskHandle, defaultPriority);
>     }
>     osDelay(1);
>   }
> }
> ```
> 
> Basically, when vscp message is received I increase task priority so it can consume all messages as fast as it can and then decrease proiorty to normal so timers/states can be processed inside _vscp_core_process()_ as usual.
> 
> _vscp_timer_process(100);_ is processed in RTOS timer callback every 100ms in separate thread.
> 
> just for the reference, I have this in _vscp_tp_adapter.c_ :
> 
> ```
> extern BOOL vscp_tp_adapter_readMessage(vscp_RxMessage *const msg) {
>   ...
>   HALP_CAN_Status_t status = HALP_CAN_ReceiveFromQueue(&hCAN, 0);
>   ...
> ```
> 
> So I don't block _vscp_core_process()_ at all... but putting _osWaitForever_ instead _0_ it would block until message is received and with current implementation we can't do that because timers could expire but won't be processed in state machine.

I see ... my hope was about osWait(), but it died unfortunately.

**Comment by BlueAndi** [comment]
I updated the vscp_core_process() return value.

**Comment by troky** [comment]
cool. thanks!
